My Simple App/Business Logic/AddressStore/AddressStoreImpl.swift

```diff
@@ -0,0 +1,123 @@
+//
+//  AddressStoreImpl.swift
+//  My Simple App
+//
+//  Created by Erick Sanchez on 8/12/20.
+//  Copyright © 2020 Erick Sanchez. All rights reserved.
+//
+
+import Foundation
+
+func injectAddressStore() -> AddressStore {
+  AddressStoreImpl(internalUserDefaults: injectInternalUserDefaults())
+}
+
+class AddressStoreImpl: AddressStore {
+
+  // TODO: mock dependencies and write unit tests.
+
+  private static let userAddressesKey = "USER_ADDRESSES"
+
+  fileprivate let storeDidChange = Event<Void>()
+
+  fileprivate let internalUserDefaults: InternalUserDefaults
+
+  init(internalUserDefaults: InternalUserDefaults) {
+    self.internalUserDefaults = internalUserDefaults
+  }
+
+  func addresses() -> AllAddressesResource {
+    return AnyReactableResource(base: AllAddressesReactableResource(store: self))
+  }
+
+  func storeAddress(name: String) {
+    let addressObject = AddressObject(
+      identifier: UUID(),
+      name: name,
+      longitude: nil,
+      latitude: nil,
+      temperature: nil,
+      iconName: nil
+    )
+    let newAddresses = [addressObject] + self.loadData()
+    self.saveData(addresses: newAddresses)
+  }
+
+  func storeAddress(name: String, latitude: Double, longitude: Double, weather: Weather) {
+    let addressObject = AddressObject(
+      identifier: UUID(),
+      name: name,
+      longitude: longitude,
+      latitude: latitude,
+      temperature: weather.temperature,
+      iconName: weather.iconName
+    )
+    let newAddresses = [addressObject] + self.loadData()
+    self.saveData(addresses: newAddresses)
+  }
+
+  func delete(address: Address) {
+    let idMatcher: (AddressObject) -> Bool = { $0.identifier == address.identifier }
+    var addressesData = self.loadData()
+
+    guard let indexToRemove = addressesData.firstIndex(where: idMatcher) else {
+      return assertionFailure("Failed to delete given address, not found in database: \(address)")
```

Comment:
woodward4422 2 days ago
Why assertionFailure here? I am not a fan of this convention

LinnierGames 42 minutes ago Author Owner
How I've seen assertions used in our codebase is to guard against programmer errors and never user errors. If the programmer error is something that should never happen and is a SUPER incorrect use of the code, then use fatalError(...). For any of these asserts, we always put a message

In this case, an ID is expected to always exist bc the AddressStoreImpl provided the Address object to the client-API and no one else should be able to instantiate an invalid Address object. Therefore, if the guard fails then something is wrong internally